tests/environment.py

```diff
@@ -0,0 +1,175 @@
+###############################################################
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+# Copyright (c) 2020 Arnout Vandecappelle (Essensium/Mind)
+# This code is subject to the terms of the BSD+Patent license.
+# See LICENSE file for more details.
+###############################################################
+
+import json
+import os
+import re
+import subprocess
+
+from capi import UCCSocket
+from opts import debug
+import sniffer
+
+
+class AlEntity:
+    '''Abstract representation of a MultiAP device (1905.1 AL Entity).
+
+    Derived classes provide concrete implementations for a specific device (e.g. docker).
+
+    This provides basic information about the entity, e.g. its AL MAC address. How this information
+    is retrieved is implementation-specific.
+
+    It also provides an abstract interface to interact with the entity, e.g. for sending CAPI
+    commands.
+
+    If a device runs both the agent and the controller, two AlEntities should be created for it,
+    with the same MAC address. This is not how it is modeled in 1905.1, but it corresponds to how
+    it is implemented in prplMesh and it allows us to have e.g. a separate UCCSocket to controller
+    and agent.
+    '''
+    def __init__(self, mac: str, ucc_socket: UCCSocket, is_controller: bool = False):
+        self.mac = mac
+        self.ucc_socket = ucc_socket
+        self.is_controller = is_controller
+        self.radios = []
+
+        # Convenience functions that propagate to ucc_socket
+        self.cmd_reply = self.ucc_socket.cmd_reply
+        self.dev_get_parameter = self.ucc_socket.dev_get_parameter
+        self.dev_send_1905 = self.ucc_socket.dev_send_1905
+
+    def command(self, *command: str) -> bytes:
+        '''Run `command` on the device and return its output as bytes.'''
+        raise NotImplementedError("command is not implemented in abstract class AlEntity")
+
+
+class Radio:
+    '''Abstract representation of a radio on a MultiAP agent.
+
+    This provides basic information about the radio, e.g. its mac address, and functionality for
+    checking its status.
+    '''
+    def __init__(self, agent: AlEntity, mac: str):
+        self.agent = agent
+        agent.radios.append(self)
+        self.mac = mac
+
+
+# The following variables are initialized as None, and have to be set when a concrete test
+# environment is started.
+wired_sniffer = None
+controller = None
+agents = []
+
+
+# Concrete implementation with docker
+
+rootdir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+installdir = os.path.join(rootdir, 'build', 'install')
+
+
+class AlEntityDocker(AlEntity):
```

Comment:
Same comment as before: I'd capitalize ALEntity, even if just to avoid confusion (nitpick)